File: src/main/resources/public/ts/services/QuestionService.ts
import {idiom, ng, notify} from 'entcore';
import http, {AxiosResponse} from 'axios';
import {Question} from "../models";

export interface QuestionService {
    list(formId: number) : Promise<AxiosResponse>;
    countQuestions(formId: number) : Promise<AxiosResponse>;
    get(questionId: number) : Promise<AxiosResponse>;
    getByPosition(idForm: number, position: number) : Promise<AxiosResponse>;
    save(question: Question) : Promise<AxiosResponse>;
    create(question: Question) : Promise<AxiosResponse>;
    createMultiple(questions: Question[], formId: number) : Promise<AxiosResponse>;
    update(question: Question) : Promise<AxiosResponse>;
    delete(questionId: number) : Promise<AxiosResponse>;
}

export const questionService: QuestionService = {

    async list (formId: number) : Promise<AxiosResponse> {
        try {
            return http.get(`/formulaire/forms/${formId}/questions`);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.questionService.list'));
            throw err;
        }
    },

    async countQuestions (formId: number) : Promise<AxiosResponse> {
        try {
            return http.get(`/formulaire/forms/${formId}/questions/count`);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.questionService.count'));
            throw err;
        }
    },

    async get(questionId: number) : Promise<AxiosResponse> {
        try {
            return http.get(`/formulaire/questions/${questionId}`);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.questionService.get'));
            throw err;
        }
    },

    async getByPosition(idForm : number, position: number) : Promise<AxiosResponse> {
        try {
            return http.get(`/formulaire/forms/${idForm}/questions/${position}`);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.questionService.get'));
            throw err;
        }
    },

    async save(question: Question) : Promise<AxiosResponse> {
        return question.id ? await this.update(question) : await this.create(question);
    },

    async create(question: Question) : Promise<AxiosResponse> {
        try {
            return http.post(`/formulaire/forms/${question.form_id}/questions`, question);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.questionService.create'));
            throw err;
        }
    },

    async createMultiple(questions: Question[], formId : number) : Promise<AxiosResponse> {
        try {
            return http.post(`/formulaire/forms/${formId}/questions/m`, questions);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.questionService.createMultiple'));
            throw err;
        }
    },

    async update(question: Question) : Promise<AxiosResponse> {
        try {
            return http.put(`/formulaire/questions/${question.id}`, question);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.questionService.update'));
            throw err;
        }
    },

    async delete(questionId: number) : Promise<AxiosResponse> {
        try {
            return http.delete(`/formulaire/questions/${questionId}`);
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.questionService.delete'));
            throw err;
        }
    }
};

export const QuestionService = ng.service('QuestionService', (): QuestionService => questionService);

File: formulaire/src/main/java/fr/openent/formulaire/service/QuestionService.java
package fr.openent.formulaire.service;

import fr.openent.form.core.models.Question;
import fr.wseduc.webutils.Either;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

import java.util.List;
import java.util.Optional;

public interface QuestionService {
    /**
     * List all the questions out of sections of a specific form
     * @param formId form identifier
     * @param handler function handler returning JsonArray data
     */
    void listForForm(String formId, Handler<Either<String, JsonArray>> handler);

    /**
     * List all the questions out of sections of a specific form
     * @param formId form identifier
     */
    Future<JsonArray> listForForm(String formId);

    /**
     * List all the questions in a specific section
     * @param sectionId section identifier
     * @param handler function handler returning JsonArray data
     */
    void listForSection(String sectionId, Handler<Either<String, JsonArray>> handler);

    /**
     * List all the questions (in and out of sections) of a specific form
     * @param formId section identifier
     */
    Future<JsonArray> listForFormAndSection(String formId);

    /**
     * List all the children questions of a list of questions
     * @param questionIds JsonArray identifier
     */
    Future<JsonArray> listChildren(JsonArray questionIds);

    /**
     * List all the children questions of a list of questions
     * @param questionIds question identifiers
     * @param handler function handler returning JsonArray data
     */
    void listChildren(JsonArray questionIds, Handler<Either<String, JsonArray>> handler);

    /**
     * List all the questions of a specific form without freetext questions
     * @param formId form identifier
     * @param isPdf is export for PDF
     * @param handler function handler returning JsonArray data
     */
    void getExportInfos(String formId, boolean isPdf, Handler<Either<String, JsonArray>> handler);

    /**
     * List all the questions of a specific form without freetext questions
     * @param formId form identifier
     * @param isPdf is export for PDF
     */
    Future<JsonArray> getExportInfos(String formId, boolean isPdf);

    /**
     * Get a specific question by id
     * @param questionId question identifier
     * @param handler function handler returning JsonObject data
     */
    void get(String questionId, Handler<Either<String, JsonObject>> handler);

    /**
     * Get the section ids with conditional questions for a specific form
     * @param formId form identifier
     * @param questionIds question identifiers to omit
     * @param handler function handler returning JsonArray data
     */
    void getSectionIdsWithConditionalQuestions(String formId, JsonArray questionIds, Handler<Either<String, JsonArray>> handler);

    /**
     * Get the section ids of a specific form thanks to a question id
     * @param questionId question identifier
     * @param handler function handler returning JsonArray data
     */
    void getSectionIdsByForm(String questionId, Handler<Either<String, JsonArray>> handler);

    /**
     * Get form position corresponding to a specific question
     * @param questionId question identifier
     * @param handler function handler returning JsonObject data
     */
    void getFormPosition(String questionId, Handler<Either<String, JsonObject>> handler);

    /**
     * Create a question in a specific form
     * @param question Question data
     * @param formId form identifier
     */
    Future<Optional<Question>> create(Question question, String formId);

    /**
     * Update specific questions
     * @param formId question identifier
     * @param questions JsonArray data
     * @param handler function handler returning JsonArray data
     */
    void update(String formId, JsonArray questions, Handler<Either<String, JsonArray>> handler);

    /**
     * Update specific questions
     * @param formId question identifier
     * @param questions JsonArray data
     */
    Future<List<Question>> update(String formId, List<Question> questions);

    /**
     * Delete a specific question
     * @param question question
     * @param handler function handler returning JsonObject data
     */
    void delete(JsonObject question, Handler<Either<String, JsonObject>> handler);
}
